src/main/scala/com/redhat/et/silex/som/som.scala

```diff
@@ -0,0 +1,218 @@
+
+    new SOM(xdim, ydim, fdim, randomMap, SampleSink.empty)
+  }
+  
+  /** Create a new SOM instance with the results of the training state */
+  private [som] def step(xdim: Int, ydim: Int, fdim: Int, state: SomTrainingState, xsigma: Double, ysigma: Double, lastState: SOM) = {
+    var weights = DenseVector.fill[DenseVector[Double]](xdim * ydim)(DenseVector.zeros[Double](fdim))
+    val seenWeights = DenseVector[DenseVector[Double]](state.weights)
+    var neighborhoods = DenseVector.zeros[Double](xdim * ydim)
+    
+    (0 until xdim * ydim).foreach { idx =>
+      val (xc, yc) = (idx / ydim, idx % ydim)
+      val counts = state.counts(idx).toDouble
+      val hood = Neighborhood.mat(xc, xdim, xsigma, yc, ydim, ysigma).reshape(xdim * ydim, 1).toDenseVector
+      neighborhoods = neighborhoods :+ (hood * counts)
+      val update = DenseVector(hood.toArray.map { hoodDistance => state.weights(idx) * hoodDistance})
+      weights = weights :+ update
+    }
+    
+    val newWeights = DenseVector((weights.values.iterator.zip(lastState.entries.iterator)).zip(neighborhoods.values.iterator).map { 
+      case ((vd, od), d) if d == 0.0 => od
+      case ((vd, _), d) => vd / d 
+    }.toArray)
+    new SOM(xdim, ydim, fdim, newWeights, state.matchQuality)
+  }
+  
+  @inline private [som] def spark2breeze(vec: SV): Vector[Double] = {
+    vec match {
+      case sv: SSV => {
+        val vb = new VectorBuilder[Double](sv.size)
+        (sv.indices zip sv.values).foreach { case (idx: Int, v: Double) =>
+          vb.add(idx, v)
+        }
+        vb.toSparseVector()
+      }
+      
+      case dv: SDV => {
+        DenseVector(dv.toArray)
+      }
+    }
+  }
+  
+  /** Train a self-organizing map from an RDD of <tt>examples</tt> for <tt>iterations</tt> iterations, starting with a randomly-weighted map with <tt>xdim</tt> columns, <tt>ydim</tt> rows, and <tt>fdim</tt> features per cell. Optional parameters include <tt>sigmaScale</tt>, to specify the width of the neighborhood function as a factor of each map dimension, <tt>minSigma</tt>, which sets the neighborhood width at the last iteration, and <tt>hook</tt>, which is a callback function to run after each iteration. */
+  def train(xdim: Int, ydim: Int, fdim: Int, iterations: Int, examples: RDD[SV], seed: Option[Int] = None, sigmaScale: Double = 0.95, minSigma: Double = 1, hook: (Int, SOM) => Unit = { case (_,_) => }): SOM = {
+    val xSigmaStep = ((xdim * sigmaScale) - minSigma) / iterations
+    val ySigmaStep = ((ydim * sigmaScale) - minSigma) / iterations
+    val sc = examples.context
+    val normedExamples = examples.map { 
+      ex => {
+        val bv = spark2breeze(ex)
+        (bv, norm(bv))
+      }
+    }.cache
+    
+    val startState = random(xdim, ydim, fdim, seed)
+    
+    hook(0, startState)
+    
+    (0 until iterations).foldLeft(startState) { (acc, it) =>
```

Comment:
Some kind of adaptive halting would be good, to avoid doing too many iterations (or too few).  Possibly some measure of small aggregate change in closest-distances over the training elements.  Another variation I like is halting when training samples are no longer shifting, i.e.  a sufficiently small fraction of them had a different closest-cell on the previous iteration.